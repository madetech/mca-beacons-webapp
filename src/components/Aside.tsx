import React, { FunctionComponent, ReactNode } from "react";
import styles from "../styles/Aside.module.scss";

interface AsideProps {
  title?: string;
  children: ReactNode;
}

const Aside: FunctionComponent<AsideProps> = ({
  title = null,
  children,
}: AsideProps): JSX.Element => (
  <>
    <aside
      className={`${styles.related_items} govuk-!-padding-top-3 govuk-!-padding-bottom-4`}
      role="complementary"
    >
      {title !== null && (
        <h2 className="govuk-heading-m" id="subsection-title">
          {title}
        </h2>
      )}
      <nav role="navigation" aria-labelledby="subsection-title">
        {children}
      </nav>
    </aside>
  </>
);

export default Aside;
