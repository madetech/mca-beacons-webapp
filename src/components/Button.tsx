import React, { FunctionComponent } from "react";
import Link from "next/link";

interface ButtonProps {
  buttonText: string;
}

interface StartButtonProps {
  buttonText?: string;
  href: string;
}

interface BackButtonProps {
  href: string;
}

export const Button: FunctionComponent<ButtonProps> = ({
  buttonText,
}: ButtonProps): JSX.Element => (
  <button className="govuk-button" data-module="govuk-button">
    {buttonText}
  </button>
);

export const StartButton: FunctionComponent<StartButtonProps> = ({
  buttonText = "Start now",
  href,
}: StartButtonProps): JSX.Element => (
  <Link href={href}>
    <a
      role="button"
      draggable="false"
      className="govuk-button govuk-button--start"
      data-module="govuk-button"
    >
      {buttonText}
      <svg
        className="govuk-button__start-icon"
        xmlns="http://www.w3.org/2000/svg"
        width="17.5"
        height="19"
        viewBox="0 0 33 40"
        aria-hidden="true"
        focusable="false"
      >
        <path fill="currentColor" d="M0 0h13l20 20-20 20H0l20-20z" />
      </svg>
    </a>
  </Link>
);

export const BackButton: FunctionComponent<BackButtonProps> = ({
  href,
}: BackButtonProps): JSX.Element => (
  <a href={href} className="govuk-back-link">
    Back
  </a>
);
