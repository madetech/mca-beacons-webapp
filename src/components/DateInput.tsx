import React, { FunctionComponent, ReactNode } from "react";

interface DateListInputProps {
  id: string;
  children: ReactNode;
}

interface DateListItemInputProps {
  children: ReactNode;
}

interface DateInputProps {
  id: string;
  dateType: DateType;
  name?: string;
  defaultValue?: string;
  className?: string;
}

export enum DateType {
  DAY,
  MONTH,
  YEAR,
}

export const DateListInput: FunctionComponent<DateListInputProps> = ({
  id,
  children,
}: DateListInputProps): JSX.Element => (
  <div className="govuk-date-input" id={id}>
    {children}
  </div>
);

export const DateListItem: FunctionComponent<DateListItemInputProps> = ({
  children,
}: DateListItemInputProps): JSX.Element => (
  <div className="govuk-date-input__item">{children}</div>
);

export const DateInput: FunctionComponent<DateInputProps> = ({
  id,
  dateType,
  name = null,
  defaultValue = "",
  className = "",
}: DateInputProps) => {
  const dateTypeClass: string =
    dateType === DateType.YEAR
      ? "govuk-input--width-4"
      : "govuk-input--width-2";

  name = name ? name : id;

  return (
    <input
      className={`govuk-input govuk-date-input__input ${dateTypeClass} ${className}`}
      id={id}
      name={name}
      type="text"
      pattern="[0-9]*"
      defaultValue={defaultValue}
      inputMode="numeric"
    />
  );
};
