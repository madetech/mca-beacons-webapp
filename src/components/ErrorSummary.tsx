import React, { FunctionComponent, ReactNode } from "react";
import { IFormError } from "../lib/formValidator";
import { toArray } from "../lib/utils";

interface FormErrorSummaryProps {
  errors: IFormError[];
  showErrors?: boolean;
}

interface FormErrorSummaryLinkProps {
  href: string;
  errorMessage: string;
}

interface FieldErrorListProps {
  errorMessages: string[];
}

interface FieldErrorMessageProps {
  errorMessage: string;
}

interface ErrorSummaryProps {
  children: ReactNode;
}

export const FormErrorSummary: FunctionComponent<FormErrorSummaryProps> = ({
  errors,
  showErrors = false,
}: FormErrorSummaryProps) => {
  errors = toArray(errors);
  let errorComponent: ReactNode;

  if (showErrors && errors.length > 0) {
    errorComponent = (
      <ErrorSummary>
        {errors.map((field) =>
          field.errorMessages.map((errorMessage, index) => (
            <FormErrorSummaryLink
              key={`${field.fieldId}-${index}`}
              href={`#${field.fieldId}`}
              errorMessage={errorMessage}
            />
          ))
        )}
      </ErrorSummary>
    );
  }

  return <>{errorComponent}</>;
};

const FormErrorSummaryLink: FunctionComponent<FormErrorSummaryLinkProps> = ({
  href,
  errorMessage,
}: FormErrorSummaryLinkProps) => (
  <li>
    <a href={href}>{errorMessage}</a>
  </li>
);

export const FieldErrorList: FunctionComponent<FieldErrorListProps> = ({
  errorMessages,
}: FieldErrorListProps): JSX.Element => (
  <>
    {errorMessages.map((message, index) => (
      <FieldErrorMessage errorMessage={message} key={`${index}`} />
    ))}
  </>
);

const FieldErrorMessage: FunctionComponent<FieldErrorMessageProps> = ({
  errorMessage,
}: FieldErrorMessageProps) => (
  <span className="govuk-error-message">
    <span className="govuk-visually-hidden">Error:</span>
    {errorMessage}
  </span>
);

const ErrorSummary: FunctionComponent<ErrorSummaryProps> = ({
  children,
}: ErrorSummaryProps): JSX.Element => (
  <div
    className="govuk-error-summary"
    aria-labelledby="error-summary-title"
    role="alert"
    tabIndex={-1}
    data-module="govuk-error-summary"
  >
    <h2 className="govuk-error-summary__title" id="error-summary-title">
      There is a problem
    </h2>
    <div className="govuk-error-summary__body">
      <ul className="govuk-list govuk-error-summary__list">{children}</ul>
    </div>
  </div>
);
