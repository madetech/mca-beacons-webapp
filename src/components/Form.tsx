import React, {
  FunctionComponent,
  InputHTMLAttributes,
  PropsWithChildren,
  ReactNode,
  SelectHTMLAttributes,
} from "react";
import { HttpMethod } from "../lib/types";

interface FormProps {
  action: string;
  method?: HttpMethod;
  children: ReactNode;
}

interface FormFieldsetProps {
  ariaDescribedBy?: string;
  children: ReactNode;
}

interface FormGroupProps {
  children: ReactNode;
  hasError?: boolean;
}

interface FormLabelProps {
  htmlFor: string;
  className?: string;
  children: ReactNode;
}

interface FormHintProps {
  forId: string;
  className?: string;
  children: ReactNode;
}

interface FormLegendProps {
  children: ReactNode;
}

interface FormLegendPageHeadingProps {
  children: ReactNode;
}

interface InputProps {
  name: string;
  id?: string;
  type?: string;
  htmlAttributes?: InputHTMLAttributes<Element>;
}

interface SelectProps {
  name: string;
  defaultValue: string;
  children: ReactNode;
  id?: string;
  htmlAttributes?: SelectHTMLAttributes<Element>;
}

interface SelectOptionProps {
  value: string;
  children: ReactNode;
}

export const Form: FunctionComponent<FormProps> = ({
  action,
  method = HttpMethod.POST,
  children,
}: FormProps): JSX.Element => (
  <form action={action} method={method}>
    {children}
  </form>
);

export const FormGroup: FunctionComponent<FormGroupProps> = ({
  children,
  hasError = false,
}: FormGroupProps): JSX.Element => (
  <div
    className={`govuk-form-group ${hasError ? "govuk-form-group--error" : ""}`}
  >
    {children}
  </div>
);

export const FormFieldset: FunctionComponent<FormFieldsetProps> = ({
  ariaDescribedBy = null,
  children,
}: PropsWithChildren<FormFieldsetProps>): JSX.Element => (
  <fieldset className="govuk-fieldset" aria-describedby={ariaDescribedBy}>
    {children}
  </fieldset>
);

export const FormLegend: FunctionComponent<FormLegendProps> = ({
  children,
}: FormLegendProps): JSX.Element => (
  <legend className="govuk-fieldset__legend">{children}</legend>
);

export const FormLegendPageHeading: FunctionComponent<FormLegendPageHeadingProps> = ({
  children,
}: FormLegendPageHeadingProps) => (
  <legend className="govuk-fieldset__legend govuk-fieldset__legend--l">
    <h1 className="govuk-fieldset__heading">{children}</h1>
  </legend>
);

export const FormLabel: FunctionComponent<FormLabelProps> = ({
  htmlFor,
  className = "",
  children,
}: FormLabelProps): JSX.Element => (
  <label className={`govuk-label ${className}`} htmlFor={htmlFor}>
    {children}
  </label>
);

export const FormHint: FunctionComponent<FormHintProps> = ({
  forId,
  className = "",
  children,
}: FormHintProps): JSX.Element => (
  <div id={`${forId}-hint`} className={`govuk-hint ${className}`}>
    {children}
  </div>
);

export const Input: FunctionComponent<InputProps> = ({
  id = null,
  name,
  type = "text",
  htmlAttributes = {},
}: InputProps): JSX.Element => (
  <input
    className="govuk-input"
    id={id}
    name={name}
    type={type}
    {...htmlAttributes}
  />
);

export const Select: FunctionComponent<SelectProps> = ({
  id = null,
  name,
  defaultValue,
  children,
  htmlAttributes = {},
}: SelectProps): JSX.Element => (
  <select
    className="govuk-select"
    id={id}
    name={name}
    defaultValue={defaultValue}
    {...htmlAttributes}
  >
    {children}
  </select>
);

export const SelectOption: FunctionComponent<SelectOptionProps> = ({
  value,
  children,
}: SelectOptionProps): JSX.Element => <option value={value}>{children}</option>;
