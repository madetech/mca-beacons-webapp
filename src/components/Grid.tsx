import React, { FunctionComponent, ReactNode } from "react";

interface GridProps {
  mainContent: ReactNode;
  aside?: ReactNode;
}

export const Grid: FunctionComponent<GridProps> = ({
  mainContent,
  aside = null,
}: GridProps) => (
  <div className="govuk-grid-row">
    <div className="govuk-grid-column-two-thirds">{mainContent}</div>

    {aside && <div className="govuk-grid-column-one-third">{aside}</div>}
  </div>
);
