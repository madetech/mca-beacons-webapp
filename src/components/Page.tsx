import React from "react";
import { Header } from "./Header";
import { Footer } from "./Footer";
import { PhaseBanner } from "./PhaseBanner";

interface PageProps {
  serviceName: string;
  homeLink: string;
  phase: string;
  bannerHtml: JSX.Element;
  children: JSX.Element[];
}

export const Page: React.FC<PageProps> = (props: PageProps): JSX.Element => (
  <React.Fragment>
    <Header homeLink={props.homeLink} serviceName={props.serviceName} />
    <PhaseBanner phase={props.phase} bannerHtml={props.bannerHtml} />
    {props.children}
    <Footer />
  </React.Fragment>
);
