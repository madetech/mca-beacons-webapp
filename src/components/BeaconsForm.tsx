import { useRouter } from "next/router";
import React, { FunctionComponent, ReactNode } from "react";
import { FormError } from "../lib/form/formManager";
import { BackButton, BackButtonRouterIndexes, Button } from "./Button";
import { FormErrorSummary } from "./ErrorSummary";
import { Form, FormFieldset, FormGroup, FormLegendPageHeading } from "./Form";
import { Grid } from "./Grid";
import { InsetText } from "./InsetText";
import { Layout } from "./Layout";
import { IfYouNeedHelp } from "./Mca";

interface BeaconsFormProps {
  children: ReactNode;
  previousPageUrl: string;
  pageHeading: string;
  showCookieBanner: boolean;
  formErrors?: FormError[];
  errorMessages?: string[];
  insetText?: ReactNode;
  includeUseIndex?: boolean;
}

export const BeaconsForm: FunctionComponent<BeaconsFormProps> = ({
  children,
  previousPageUrl,
  pageHeading,
  showCookieBanner,
  formErrors = [],
  errorMessages = [],
  insetText = null,
  includeUseIndex = false,
}: BeaconsFormProps): JSX.Element => {
  let insetComponent: ReactNode;
  if (insetText) {
    insetComponent = <InsetText>{insetText}</InsetText>;
  }

  const backButton: ReactNode = includeUseIndex ? (
    <BackButtonRouterIndexes href={previousPageUrl} />
  ) : (
    <BackButton href={previousPageUrl} />
  );

  return (
    <Layout
      navigation={backButton}
      title={pageHeading}
      showCookieBanner={showCookieBanner}
    >
      <Grid
        mainContent={
          <>
            <FormErrorSummary formErrors={formErrors} />
            <Form>
              <FormGroup errorMessages={errorMessages}>
                <FormFieldset>
                  <FormLegendPageHeading>{pageHeading}</FormLegendPageHeading>
                </FormFieldset>
                {insetComponent}
                {children}
                <HiddenInput />
              </FormGroup>
              <Button buttonText="Continue" />
            </Form>
            <IfYouNeedHelp />
          </>
        }
      />
    </Layout>
  );
};

const HiddenInput: FunctionComponent = () => {
  const router = useRouter();
  const useIndexValue = router.query.useIndex || 0;

  return (
    <input id="use-index" type="hidden" name="useIndex" value={useIndexValue} />
  );
};
