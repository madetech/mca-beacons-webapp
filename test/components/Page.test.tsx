import { render } from "@testing-library/react";
import React from "react";
import { Page } from "../../src/components/Page";

describe("Page", () => {
  it("renders the Page correctly", () => {
    const pageProps = {
      serviceName: "Beacons Beacons Beacons",
      homeLink: "#",
      phase: "BETA",
      bannerHtml: <>This is just a test service, please ignore.</>,
    };
    const { asFragment } = render(
      <Page {...pageProps}>
        <h1>Welcome to Beacons Beacons Beacons!</h1>
        <p>For all your beacon and beacon accessory needs.</p>
      </Page>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
