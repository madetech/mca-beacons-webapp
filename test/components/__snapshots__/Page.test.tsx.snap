// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page renders the Page correctly 1`] = `<DocumentFragment />`;
